### PR TITLE
Parallel move and copy commands

### PR DESCRIPTION
**Pull request by barbelity**
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: artifactory/services/utils/searchutil.go [hunk 1]
```diff
@@ -267,22 +286,35 @@ func (item *ResultItem) ToDependency() buildinfo.Dependency {
...
+func FilterBottomChainResults(readerRecord SearchBasedContentItem, reader *content.ContentReader) (*content.ContentReader, error) {
 	writer, err := content.NewContentWriter(content.DefaultKey, true, false)
 	if err != nil {
 		return nil, err
 	}
 	defer writer.Close()
+
+	// Get the expected record type from the reader.
+	recordValue := reflect.ValueOf(readerRecord)
+	recordType := recordValue.Type()
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/utils/searchutil.go:300]: Fixed.

File: artifactory/services/movecopy.go [hunk 2]
```diff
@@ -49,24 +51,40 @@ func (mc *MoveCopyService) GetJfrogHttpClient() (*rthttpclient.ArtifactoryHttpCl
...
 		return 0, 0, err
 	}
-	defer resultReader.Close()
-	successCount, failedCount, err = mc.moveFiles(resultReader, moveSpec)
+	defer tempAggregatedReader.Close()
+
+	aggregatedReader := tempAggregatedReader
+	if mc.moveType == MOVE {
+		// If move command, reduce top dir chain results.
+		defer tempAggregatedReader.Close()
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:79]: Removed the redundant Close().

File: artifactory/services/movecopy.go [hunk 3]
```diff
@@ -49,24 +51,40 @@ func (mc *MoveCopyService) GetJfrogHttpClient() (*rthttpclient.ArtifactoryHttpCl
...
+func (mc *MoveCopyService) MoveCopyServiceMoveFilesWrapper(moveSpecs ...MoveCopyParams) (successCount, failedCount int, err error) {
+	moveReaders := []*ReaderSpecTuple{}
+	defer func() {
+		for _, readerSpec := range moveReaders {
+			readerSpec.Reader.Close()
+		}
+	}()
+	for i, moveSpec := range moveSpecs {
+		// Create reader for each spec.
+		moveReader, err := mc.getPathsToMove(moveSpec)
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:63]: All moveReaders are being closed by the defer func() few lines above.

File: artifactory/services/movecopy.go [hunk 4]
```diff
@@ -154,43 +243,76 @@ func (mc *MoveCopyService) moveFile(sourcePath, destPath string) (bool, error) {
...
-		log.Error("Artifactory response: " + resp.Status + "\n" + clientutils.IndentJson(body))
+		log.Error(logMsgPrefix + "Artifactory response: " + resp.Status + "\n" + clientutils.IndentJson(body))
 	}
 
-	log.Debug("Artifactory response:", resp.Status)
+	log.Debug(logMsgPrefix+"Artifactory response:", resp.Status)
 	return resp.StatusCode == http.StatusOK, nil
 }
 
+// Receive multiple 'ReaderSpecTuple' items and merge them into a single 'ContentReader' of 'MoveResultItem'.
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:283]: Changed.

File: artifactory/services/movecopy.go [hunk 5]
```diff
@@ -154,43 +243,76 @@ func (mc *MoveCopyService) moveFile(sourcePath, destPath string) (bool, error) {
...
+		}
+		if err := cr.GetError(); err != nil {
+			return nil, err
+		}
+	}
+	return content.NewContentReader(cw.GetFilePath(), arrayKey), nil
+}
+
+func promptMoveCopyMessage(reader *content.ContentReader, moveType MoveType) {
+	length, err := reader.Length()
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:305]: The length is being used for logging how many artifacts are actually going to be moved/copied.
Should we remove this log message?

File: artifactory/services/movecopy.go [hunk 6]
```diff
@@ -75,72 +93,143 @@ func (mc *MoveCopyService) MoveCopyServiceMoveFilesWrapper(moveSpec MoveCopyPara
...
+}
+
+func reduceMovePaths(readerItem utils.SearchBasedContentItem, cr *content.ContentReader, flat bool) (*content.ContentReader, error) {
+	if flat {
+		return utils.ReduceBottomChainDirResult(readerItem, cr)
+	}
+	return utils.ReduceTopChainDirResult(readerItem, cr)
+}
+
+func (mc *MoveCopyService) parallelMoveFiles(reader *content.ContentReader, params []MoveCopyParams) (successCount, failedCount int, err error) {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:137]: Changed.

File: artifactory/services/movecopy.go [hunk 7]
```diff
@@ -75,72 +93,143 @@ func (mc *MoveCopyService) MoveCopyServiceMoveFilesWrapper(moveSpec MoveCopyPara
...
+	consumer.Run()
+	err = errorsQueue.GetError()
+	totalSuccess = utils.SumIntArray(result.SuccessCount)
+	totalFails = utils.SumIntArray(result.TotalCount) - totalSuccess
+	return
+}
+
+type fileMoveCopyHandlerFunc func(utils.ResultItem, *MoveCopyParams) parallel.TaskFunc
+
+func (mc *MoveCopyService) createFileMoveCopyHandlerFunc(result *utils.Result) fileMoveCopyHandlerFunc {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:166]: Changed.

File: artifactory/services/movecopy.go [hunk 8]
```diff
@@ -75,72 +93,143 @@ func (mc *MoveCopyService) MoveCopyServiceMoveFilesWrapper(moveSpec MoveCopyPara
...
+		log.Error(err)
+		return "", err
 	}
-	return
+
+	return destFile, nil
 }
 
-func (mc *MoveCopyService) moveFile(sourcePath, destPath string) (bool, error) {
+func (mc *MoveCopyService) moveFile(sourcePath, destPath, logMsgPrefix string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:223]: Changed.

File: artifactory/services/movecopy.go [hunk 9]
```diff
@@ -49,24 +51,40 @@ func (mc *MoveCopyService) GetJfrogHttpClient() (*rthttpclient.ArtifactoryHttpCl
...
-
-	switch moveSpec.GetSpecType() {
-	case utils.BUILD:
-		resultReader, err = utils.SearchBySpecWithBuild(moveSpec.GetFile(), mc)
-	case utils.AQL:
-		resultReader, err = utils.SearchBySpecWithAql(moveSpec.GetFile(), mc, utils.NONE)
-	case utils.WILDCARD:
-		moveSpec.SetIncludeDir(true)
-		resultReader, err = utils.SearchBySpecWithPattern(moveSpec.GetFile(), mc, utils.NONE)
+func (mc *MoveCopyService) MoveCopyServiceMoveFilesWrapper(moveSpecs ...MoveCopyParams) (successCount, failedCount int, err error) {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:54]: Changed.

File: artifactory/services/movecopy.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (eyalbe4) [reply artifactory/services/movecopy.go]: OK.

File: artifactory/services/movecopy.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (eyalbe4) [reply artifactory/services/movecopy.go:306]: Let's keep the log message, but avoid calling the Length() method. It doesn't really matters if the log message includes ```artifact``` or ```artifacts```. Let's have always ```artifacts``` regardless of the length.

[the remark this answers is not in the document]
Comment (barbelity) [reply artifactory/services/movecopy.go:306]: We agreed to keep the log printing.
Opening https://github.com/jfrog/jfrog-client-go/issues/252 for improving the Length() performance.

**Pull request merged**